### PR TITLE
Add support for using a different MGF1 hash with RSA-OAEP

### PR DESCRIPTION
**Pull request by samitolvanen**
OP-TEE currently doesn't support using a different hash algorithm for MGF1 with RSA-OAEP. However, this is required for AOSP compatibility (e.g. in the [EncryptionOperationsTest.RsaOaepWithMGFDigestSuccess](https://android.googlesource.com/platform/hardware/interfaces/+/refs/heads/main/security/keymint/aidl/vts/functional/KeyMintTest.cpp#5552) KeyMint VTS test).

This PR cherry-picks the upstream libtomcrypt commit [63091c9e5c77](https://github.com/libtom/libtomcrypt/commit/63091c9e5c776ec1c5a7a5d2973395a24e82b3ba) ("Add possibility to use different hash algorithms in RSAES-OAEP") and passes the MGF1 hash from the `TEE_ATTR_RSA_OAEP_MGF_HASH` attribute to crypto implementations.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Thanks for the pull request. I'd rather step up LTC than cherry-pick a single commit, but doing that is a bit complicated so I'll do it if @jforissier doesn't beat me to it after the upcoming OP-TEE release. For the sake of this review, we can ignore "core: ltc: Add possibility to use different hash algorithms in RSAES-AEP" and assume it will be available upstream.

It would be nice with a test in xtest to see that this continues to work.

**Comment by jenswi-linaro** [comment]
Please squash the "core: pass TEE_ATTR_RSA_OAEP_MGF_HASH to RSA-OAEP implementations"  fixup commits.

**Comment by jenswi-linaro** [comment]
> > It would be nice with a test in xtest to see that this continues to work.
> 
> I can take a look. Any preferences where this should go?

It should be in the 4xxx range of tests, either as a new case or by extending an already present case. The latter might be the easiest.

**Comment by jenswi-linaro** [comment]
I've started updating LTC with https://github.com/OP-TEE/optee_os/pull/6786. Once that's merged I'll create another PR to update LTC to the latest.

**Review by etienne-lms — COMMENTED** [review]
For commit "core: ltc: Add possibility to use different hash algorithms in RSAES-OAEP":
I think the reference to the libtomcrypt commit should be a `Link:` tag than a raw commit SHA1 that is rather expected to be a ref in the current Git revision, e.g.:
```
Link: https://github.com/libtom/libtomcrypt/commit/63091c9e5c776ec1c5a7a5d2973395a24e82b3ba`
Signed-off-by: Steffen Jaeckel <s@jaeckel.eu>
Signed-off-by: Sami Tolvanen <samitolvanen@google.com>
```
Otherwise, 
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for this commit.

Minor comments for commit "core: pass TEE_ATTR_RSA_OAEP_MGF_HASH to RSA-OAEP implementations".

File: core/lib/libtomcrypt/rsa.c [hunk 1]
```diff
@@ -356,6 +366,15 @@ TEE_Result sw_crypto_acipher_rsaes_decrypt(uint32_t algo,
 		EMSG("tee_algo_to_ltc_hashindex() returned %d", (int)res);
 		goto out;
 	}
+	if (algo != TEE_ALG_RSAES_PKCS1_V1_5) {
+		res = tee_algo_to_ltc_hashindex(mgf_algo, &ltc_mgfindex);
+		if (res != TEE_SUCCESS) {
+			EMSG("tee_algo_to_ltc_hashindex() returned %d for mgf algo %x", (int)res, mgf_algo);
```
[the remark this answers is not in the document]
Comment (samitolvanen) [reply core/lib/libtomcrypt/rsa.c]: Done.

File: lib/libmbedtls/core/rsa.c [hunk 2]
```diff
@@ -495,7 +497,9 @@ TEE_Result sw_crypto_acipher_rsaes_decrypt(uint32_t algo,
 	 */
 	if (algo != TEE_ALG_RSAES_PKCS1_V1_5) {
 		md_algo = tee_algo_to_mbedtls_hash_algo(algo);
-		if (md_algo == MBEDTLS_MD_NONE) {
+		/* Using a different MGF1 algorithm is not supported. */
+		if (md_algo == MBEDTLS_MD_NONE ||
+		    md_algo != tee_algo_to_mbedtls_hash_algo(mgf_algo)) {
```
[the remark this answers is not in the document]
Comment (samitolvanen) [reply lib/libmbedtls/core/rsa.c]: Yes, that makes sense. Done.

**Comment by samitolvanen** [comment]
> For commit "core: ltc: Add possibility to use different hash algorithms in RSAES-OAEP"

I updated the commit message for now, but I believe Jens plans to update LTC separately, and at that point I can just drop this commit from the PR.

**Comment by jforissier** [comment]
> > For commit "core: ltc: Add possibility to use different hash algorithms in RSAES-OAEP"
> 
> I updated the commit message for now, but I believe Jens plans to update LTC separately, 

Done in #6820 (merged)

> and at that point I can just drop this commit from the PR.

Please do, thanks!

File: core/lib/libtomcrypt/rsa.c [hunk 3]
```diff
@@ -356,6 +366,16 @@ TEE_Result sw_crypto_acipher_rsaes_decrypt(uint32_t algo,
 		EMSG("tee_algo_to_ltc_hashindex() returned %d", (int)res);
 		goto out;
 	}
+	if (algo != TEE_ALG_RSAES_PKCS1_V1_5) {
+		res = tee_algo_to_ltc_hashindex(mgf_algo, &ltc_mgfindex);
+		if (res != TEE_SUCCESS) {
+			EMSG("tee_algo_to_ltc_hashindex() returned %d for mgf algo %#"PRIx32,
+				(int)res, mgf_algo);
```
Comment (jenswi-linaro) [inline core/lib/libtomcrypt/rsa.c:373]: `res` should also be printed as a PRIx32.

I see now that this is copied from above, but two wrongs don't make one right.

Comment (samitolvanen) [reply core/lib/libtomcrypt/rsa.c]: Ack, also changed the message above since I'm here.

**Pull request merged**